bots/tests-data

```diff
@@ -0,0 +1,254 @@
+                    # print 'HREF', url
+                    self.hrefs.append(url)
+
+# Retrieves the content of the given URL
+def retrieve(url):
+    return urllib2.urlopen(url).read()
+
+# Returns a list of all results at the given URL
+def links(url):
+    result = [ ]
+    parser = HrefParser(url, result)
+    try:
+        parser.feed(retrieve(url))
+    except urllib2.HTTPError, ex:
+        if ex.code != 404:
+            raise
+    except urllib2.URLError:
+        pass
+    return result
+
+# Generates full GitHub JSON closed pull request
+# objects since the given date. We only care about closed
+# pull requests for processing, the rest are still in flux.
+def pulls(state='closed', since=None):
+    for pull in task.api.pulls(state=state, since=since):
+        yield pull
+
+# Generates revisions for a given pull request. Each revision
+# is a simple string sha. The first one is the one that is at
+# the head of the pull request, and any other follow on revisions
+# are those which were previously tested (eg: force push, or with
+# later fixups).
+def revisions(pull):
+    head = pull["head"]["sha"]
+    if not head:
+        return
+
+    # First give back the main pull request
+    yield head
+
+    # All the revisions we've seen
+    seen = { head: True }
+
+    # Ignore these sink urls. See below. This is an optimization.
+    # We don't need to retrieve the sink /status page for the head revision
+    ignore = { }
+
+    # Seed the set of sinks. We use these sinks to figure out additional
+    # revisions for the pull request. Unfortunately GitHub doesn't help us
+    # with a list of revisions that this pull request used to reflect. So
+    # we have to look to our sink for that info.
+    data = task.api.get("commits/{0}/status?page=1&per_page=100".format(head))
+    for status in data.get("statuses", [ ]):
+        url = status["target_url"]
+        if url:
+            ignore[urlparse.urljoin(url, "./")] = True
+            sink = urlparse.urljoin(url, "../")
+            if sink not in SINKS:
+                SINKS[sink] = links(sink)
+
+    # Now ask each sink for its set of urls
+    name = "pull-{0}".format(pull["number"])
+    for sink in SINKS:
+        for link in SINKS[sink]:
+            if link in ignore:
+                continue
+
+            # We only care about stuff at the sink where pull-XXXX is in
+            # the URL. This is how we figure out whether things are related
+            if name not in link:
+                continue
+
+            # Build a URL for the cockpituous sink /status file and read it
+            target = urlparse.urljoin(link, "status")
+            try:
+                data = json.loads(retrieve(target))
+            except ValueError, ex:
+                # print "JSON", ex
+                pass
+            except urllib2.HTTPError, ex:
+                if ex.code != 404:
+                    raise
+            except urllib2.URLError:
```

Comment:
Would it make sense to log these? Or are there so many "expected" errors here that it would just be noise?

Reply:
Added debugging lines